### ISSUE TITLE
Consider reverting the renaming of compiled c test modules (part of the pytest move)?

### ISSUE THREAD

**Issue by lesteve** [numpy/numpy#10845]
The renaming of compiled c test modules in #10790 broke the scikit-learn tests because we were using `numpy.core.umath_tests.inner1d`. There was no good reason to use `inner1d` and we now used standard numpy constructs in scikit-learn master. The problem is that it will affect the latest scikit-learn (0.19.1) as soon as numpy 1.15 is released. Given the current rate of release, it is rather likely that numpy 1.15 will be released before the next scikit-learn release (0.20).

It feels like other projects could be using some of the renamed modules, so maybe the safest approach would be to keep the old modules around for a deprecation cycle with a DeprecationWarning saying that this modules are for numpy internal use only. Would you be willing to consider this option?

I first commented in https://github.com/numpy/numpy/pull/10790#issuecomment-378285426 but @charris suggested I opened a new issue, so here it is!

**Comment by rth** [comment]
> The problem is that it will affect the latest scikit-learn (0.19.1) as soon as numpy 1.15 is released.

By affect, @lesteve means that it will raise an `ImportError` if one imports anything from `sklearn.ensemble` (cf. https://github.com/scikit-learn/scikit-learn/pull/10880#issuecomment-378245618 ) which will break existing code for a number of scikit-learn users...

**Comment by pv** [comment]
I guess
```
# numpy/core/umath_tests.py
from __future__ import division, absolute_import, print_function

import warnings
warnings.warn(("numpy.core.umath_tests is an internal Numpy "
               "module and should not be imported. It will "
               "be removed in a future Numpy release."),
              category=DeprecationWarning)

from ._umath_tests import *
```

**Comment by charris** [comment]
I'll put a shim in for this as suggested by @pv, also change the warnings in the `numpy.testing` shims to `DeprecationWarning`. For the other name changes, I will hold off on shims until there is a reported problem. An error gets noticed ...

**Comment by charris** [comment]
> Is it just a sense of "this is private, should have an _" that drives the change

Yes. I note that many of the linalg functions are using gufuncs, what exactly do you need for test cases?

**Comment by mhvk** [comment]
Functions that are simple and can be easily used to check `__array_ufunc__` works...  But I don't have a strong objection to making it `_umath_tests`; I just don't see a strong need, since most people should never look under `core` anyway (that name already makes obvious that whatever is important will be imported a level up).

**Comment by charris** [comment]
> since most people should never look under core anyway

Just so. And if it was obvious we wouldn't be having this problem :) Where are you using them for test cases? NumPy using them is not a problem.

**Comment by mhvk** [comment]
I'm not using them on anything public - my comment was mainly to support @lesteve's note that there may well be more code that expects to find these functions, and one should wonder whether the arguably minor costs are worth what seems a very minor benefit of the name change. (But since it is all about very minor things, and it doesn't really affect me, just decide what you think best!)

**Comment by charris** [comment]
It's not a minor cost.  If we change or rearrange numpy internals and people complain, that is a major cost. And we will be doing a *lot* of rearrangement going forward. If you have tests for `__array_ufunc__` shouldn't they be in NumPy? Or are they already.

**Comment by mhvk** [comment]
Yes, there are tests for `__array_ufunc__` that use these routines in numpy (but as you say, those are not a problem, we can just adjust them). I don't have anything for which the name change is an issue myself; I really am just quite aware of their existence as simple examples of `gufunc`.

Note that I am probably missing the big picture here about the larger re-arrangements that are planned.

**Comment by charris** [comment]
Ah, OK. Those tests were already adjusted. We will be merging the umath and multiarray modules. In general, I think we are heading to fewer modules in `numpy.core`.

**Comment by mhvk** [comment]
Yes, of course! That overall plan definitely sounds good!

**Comment by charris** [comment]
See #10850.

**Comment by hamogu** [comment]
I want to note that `inner1d` from `numpy.core.umath_tests` outperforms other, non-private numpy functions by a factor of 2 for row-wise dot products: 
https://stackoverflow.com/questions/15616742/vectorized-way-of-calculating-row-wise-dot-product-two-matrices-with-scipy#answer-39657770

I'm working with a numpy-based ray-tracing code where row-wise dot products are by far the most common operation that I do and I've been starting to use `numpy.core.umath_tests.inner1d` for that reason; taking the risk that you move it. There are other related Stackoverflow questions (e.g. https://stackoverflow.com/questions/6229519/numpy-column-wise-dot-product or  https://stackoverflow.com/questions/48066890/using-numpy-einsum-to-compute-inner-product-of-column-vectors-of-a-matrix) indicating some interest in a function like this.

Should I open a feature request to make `inner1d` part of the public API?

**Comment by charris** [comment]
I think we will be adding several new ufuncs for these types of stacked arrays/vectors in the future, so I would prefer to wait on that.

**Comment by hamogu** [comment]
@charris : Thanks for the fast answer. Looking forward to seeing them!

**Comment by a-recknagel** [comment]
@charris Is there a specific reason why you gave the warnings a stacklevel of 2? I don't see it turning up in the discussion, and since it isn't called in a wrapper the need isn't obvious to me. 

I noticed it because the stackframe-nesting leads to weird side-effects like pytest not being able to suppress them because it only checks the first layer.

**Comment by charris** [comment]
We try to make the error messages come from the point of call. How are you suppressing the warnings in pytest.

You might open an issue for discussion.

**Comment by a-recknagel** [comment]
[pytest does it on its own by default](https://docs.pytest.org/en/latest/warnings.html#warnings-capture). I'll open an issue then.
